### PR TITLE
Speed up isBinary() check

### PR DESCRIPTION
**Pull request by mdoube**
Remove high level things from inner loop like Object creation and
HashSet.add(), replacing them with low-level comparison logic.

If needed the same could be done using Object.equals() or similar, to
make isBinary() more generic.

Partially addresses #108 (see also https://github.com/imagej/imagej-ops/commit/dd5a3e3015e9f72e936447ba95c0cac84e822a4d#commitcomment-33284909)

### CODE REVIEW

File: Modern/utilities/src/main/java/org/bonej/utilities/ElementUtil.java [hunk 1]
```diff
@@ -119,16 +116,33 @@ private ElementUtil() {}
...
-			final T e = cursor.get();
-			colors.add(e.getRealDouble());
-			if (colors.size() > 2) {
-				return false;
+			c = cursor.get().getRealDouble();
+			//if we encounter a different pixel value
+			// than a, assign it to b, and stop
+			if (c != a) {
+				b = c;
+				break;
```
Comment (rimadoma) [inline Modern/utilities/src/main/java/org/bonej/utilities/ElementUtil.java]: Could have just one loop with

```java
double colors = 1;
while (cursor.hasNext()) {
    c = cursor.get().getRealDouble();
    if (c == a) {
        continue;
    }
    if (colors == 1) {
         b = c;
         colors++;
    } else if (c != b) {
        return false;
    }
}
```

Comment (mdoube) [reply Modern/utilities/src/main/java/org/bonej/utilities/ElementUtil.java]: true, but it means that for many if not most of the pixels you have to do an extra if, assignment and increment, unnecessarily.

File: Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java [hunk 2]
```diff
@@ -175,14 +175,14 @@ public void testIsColorsBinary() {
...
 		assertTrue("A BitType image should be binary", isBinary);
 	}
 
 	@Test(expected = NullPointerException.class)
 	public void testIsColorsBinaryThrowsNPEIfIntervalNull() {
-		final boolean result = ElementUtil.isColorsBinary(null);
+		final boolean result = ElementUtil.isBinary(null);
 
 		assertFalse("A null interval should not be binary color", result);
 	}
```
Comment (alessandrofelder) [inline Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java]: I approve the change in method name for clarity and brevity.
In line with the change in method name, the assertion string should lose the "color"?

File: Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java [hunk 3]
```diff
@@ -195,7 +195,7 @@ public void testIsColorsBinaryReturnsFalseForMulticolor() {
 			.iterator();
 		interval.cursor().forEachRemaining(e -> e.setReal(intIterator.next()));
 
-		final boolean result = ElementUtil.isColorsBinary(interval);
+		final boolean result = ElementUtil.isBinary(interval);
 
 		assertFalse(
 			"An image with more than two colours should not be binary color", result);
```
Comment (alessandrofelder) [inline Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java]: see above

File: Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java [hunk 4]
```diff
@@ -205,7 +205,7 @@ public void testIsColorsBinaryReturnsFalseForMulticolor() {
 	public void testIsColorsBinaryReturnsFalseIfIntervalEmpty() {
 		final IterableInterval<DoubleType> interval = ArrayImgs.doubles(0);
 
-		final boolean result = ElementUtil.isColorsBinary(interval);
+		final boolean result = ElementUtil.isBinary(interval);
 
 		assertFalse("An empty image should not be binary color", result);
```
Comment (alessandrofelder) [inline Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java]: see above

File: Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java [hunk 5]
```diff
@@ -214,7 +214,7 @@ public void testIsColorsBinaryReturnsFalseIfIntervalEmpty() {
 	public void testIsColorsBinaryReturnsTrueForMonochrome() {
 		final IterableInterval<DoubleType> interval = ArrayImgs.doubles(2, 2);
 
-		final boolean result = ElementUtil.isColorsBinary(interval);
+		final boolean result = ElementUtil.isBinary(interval);
 
 		assertTrue("Monochrome image should be binary color", result);
 	}
```
Comment (alessandrofelder) [inline Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java]: see above

Comment (mdoube) [reply Modern/utilities/src/test/java/org/bonej/utilities/ElementUtilTest.java]: Good catch. I've removed references to colour from the tests.

**Pull request merged**